File: src/lib/commands.ts
// Commands 封装：类型安全的 invoke 调用
// 对应 Rust 侧 commands.rs 的 11 个命令

import { invoke } from '@tauri-apps/api/core';
import { WebviewWindow } from '@tauri-apps/api/webviewWindow';
import { availableMonitors } from '@tauri-apps/api/window';
import type { Config, ImportResult, Note, NoteColor, UpdateNoteFields } from '../types/note';

/** 创建便签 */
export function createNote(params: {
  title?: string;
  content?: string;
  color?: NoteColor;
}): Promise<Note> {
  return invoke<Note>('create_note', {
    title: params.title ?? null,
    content: params.content ?? null,
    color: params.color ?? null,
  });
}

/** 更新便签（部分字段） */
export function updateNote(id: string, fields: UpdateNoteFields): Promise<Note> {
  return invoke<Note>('update_note', { id, fields });
}

/** 删除便签 */
export function deleteNote(id: string): Promise<void> {
  return invoke<void>('delete_note', { id });
}

/** 获取所有便签 */
export function getNotes(): Promise<Note[]> {
  return invoke<Note[]>('get_notes');
}

/** 贴出便签（更新 pinned 状态 + 前端创建便签窗口）
 * 窗口创建由前端 WebviewWindow API 完成，因 Rust 侧 WebviewUrl::App
 * 对含 query string 的路径在 Windows 上解析失败（PathBuf 不支持 ?）
 *
 * 位置策略：统一使用「全局逻辑坐标」（note.x/y 存的就是逻辑像素），
 *   WebviewWindow 的 x/y 直接传逻辑坐标即可，Tauri 会处理跨显示器和 DPI。
 *   若记录了 monitor 且坐标出界，则回退到该显示器的 (100, 100) 位置兜底。
 *   alwaysOnTop 值从 config 实时读取（Rust get_config 返回），保证与设置一致。
 */
export async function pinNote(id: string): Promise<void> {
  // 先调 Rust 更新 pinned 状态，返回 note 数据用于设置窗口初始位置尺寸
  const note = await invoke<Note>('pin_note', { id });
  // 同时获取 config.always_on_top，保证与设置一致
  let alwaysOnTop = true;
  try {
    const cfg = await invoke<Config>('get_config');
    alwaysOnTop = cfg.always_on_top ?? true;
  } catch {
    /* ignore，默认 true */
  }

  // 前端创建便签窗口，用 note 的位置和尺寸恢复
  const label = `note-${id}`;
  const url = `note.html?id=${id}`;

  // 统一用全局逻辑坐标（note.x/y 存的就是逻辑像素）
  let x = note.x;
  let y = note.y;

  // monitor 兜底：若记录了显示器，检查当前坐标是否落在所有显示器内，
  // 若出界则回退到目标显示器的 (100, 100) 逻辑位置，避免跑到屏幕外
  if (note.monitor) {
    try {
      const monitors = await availableMonitors();
      // 判断 (x, y) 逻辑坐标是否在任一显示器的逻辑矩形内
      const inAnyMonitor = monitors.some((m) => {
        const mx = Math.round(m.position.x / m.scaleFactor);
        const my = Math.round(m.position.y / m.scaleFactor);
        const mw = Math.round(m.size.width / m.scaleFactor);
        const mh = Math.round(m.size.height / m.scaleFactor);
        return x >= mx && x <= mx + mw && y >= my && y <= my + mh;
      });
      if (!inAnyMonitor) {
        const target = monitors.find((m) => (m.name ?? '') === note.monitor) ?? monitors[0];
        if (target) {
          const mx = Math.round(target.position.x / target.scaleFactor);
          const my = Math.round(target.position.y / target.scaleFactor);
          x = mx + 100;
          y = my + 100;
        }
      }
    } catch {
      /* ignore，回退到 note.x/y */
    }
  }

  const win = new WebviewWindow(label, {
    url,
    title: 'DeskNote',
    decorations: false,
    transparent: true,
    alwaysOnTop,
    skipTaskbar: true,
    resizable: true,
    visible: false, // 延迟显示，等 NoteApp 加载完数据后 show，避免空白闪烁
    width: note.width,
    height: note.height,
    x,
    y,
  });

  return new Promise((resolve, reject) => {
    win.once('tauri://created', () => resolve());
    win.once('tauri://error', (e) => reject(e));
  });
}

/** 收回便签（关闭便签窗口） */
export async function unpinNote(id: string): Promise<void> {
  await invoke<void>('unpin_note', { id });
  // 前端关闭窗口（Rust 侧也会尝试关闭，双保险）
  const label = `note-${id}`;
  try {
    const { WebviewWindow } = await import('@tauri-apps/api/webviewWindow');
    const existing = await WebviewWindow.getByLabel(label);
    if (existing) await existing.close();
  } catch {
    // 窗口可能已被 Rust 侧关闭，忽略
  }
}

/** 获取配置 */
export function getConfig(): Promise<Config> {
  return invoke<Config>('get_config');
}

/** 更新配置（部分字段，传 JSON 对象） */
export function updateConfig(partial: Partial<Config>): Promise<Config> {
  return invoke<Config>('update_config', { partial });
}

/** 导出便签到指定文件路径 */
export function exportNotes(path: string): Promise<void> {
  return invoke<void>('export_notes', { path });
}

/** 从指定文件路径导入便签 */
export function importNotes(path: string): Promise<ImportResult> {
  return invoke<ImportResult>('import_notes', { path });
}

/** 获取数据目录路径（只读，用于设置页显示） */
export function getDataDir(): Promise<string> {
  return invoke<string>('get_data_dir');
}

/** 重新注册全局快捷键（SOP 8.9，设置页改快捷键后调用） */
export function registerHotkey(): Promise<void> {
  return invoke<void>('register_hotkey');
}

/** 设置开机自启（v1.1 优化阶段 3）
 * 同时更新 config.auto_start 和系统注册项，保证两者一致
 * Rust 侧 set_auto_start 会先 update_config 再调 autolaunch.enable/disable
 */
export function setAutoStart(enabled: boolean): Promise<void> {
  return invoke<void>('set_auto_start', { enabled });
}

File: src-tauri/src/commands.rs
// DeskNote Tauri Commands
// PRD 12.6: 前端 → Rust command（增删改查），Rust → 所有窗口（event 通知）

use crate::error::{Error, Result};
use crate::events::{event_name, NotePayload};
use crate::state::{AppState, ImportResult, UpdateNoteFields};
use crate::types::{Config, Note, NoteColor};
use std::path::PathBuf;
use tauri::{AppHandle, Emitter, Manager, State};
use tauri_plugin_global_shortcut::GlobalShortcutExt;

/// SOP 10.7：落盘失败时 emit storage:error 事件，前端显示 toast
/// 命令层调用：state 写操作返回 Err 时，emit 事件后返回错误
fn emit_storage_error(app: &AppHandle, e: &Error) {
    let _ = app.emit(event_name::STORAGE_ERROR, e.to_string());
}

/// 创建便签
/// color 为 None 时使用 config.default_color（PRD 设置项 3）
#[tauri::command]
pub fn create_note(
    app: AppHandle,
    state: State<AppState>,
    title: Option<String>,
    content: Option<String>,
    color: Option<NoteColor>,
) -> Result<Note> {
    let color = color.unwrap_or_else(|| state.get_config().default_color);
    match state.create_note(title.unwrap_or_default(), content.unwrap_or_default(), color) {
        Ok(note) => {
            let _ = app.emit(event_name::NOTE_CREATED, NotePayload { id: note.id.clone() });
            Ok(note)
        }
        Err(e) => {
            emit_storage_error(&app, &e);
            Err(e)
        }
    }
}

/// 更新便签（部分字段）
#[tauri::command]
pub fn update_note(
    app: AppHandle,
    state: State<AppState>,
    id: String,
    fields: UpdateNoteFields,
) -> Result<Note> {
    match state.update_note(&id, &fields) {
        Ok(note) => {
            let _ = app.emit(event_name::NOTE_UPDATED, NotePayload { id: note.id.clone() });
            Ok(note)
        }
        Err(e) => {
            emit_storage_error(&app, &e);
            Err(e)
        }
    }
}

/// 删除便签
#[tauri::command]
pub fn delete_note(app: AppHandle, state: State<AppState>, id: String) -> Result<()> {
    match state.delete_note(&id) {
        Ok(()) => {
            let _ = app.emit(event_name::NOTE_DELETED, NotePayload { id });
            Ok(())
        }
        Err(e) => {
            emit_storage_error(&app, &e);
            Err(e)
        }
    }
}

/// 获取所有便签
#[tauri::command]
pub fn get_notes(state: State<AppState>) -> Result<Vec<Note>> {
    Ok(state.list_notes())
}

/// 贴出便签（仅更新 pinned 状态，窗口由前端 WebviewWindow API 创建）
/// Rust 侧 WebviewUrl::App 对含 query 的路径在 Windows 上解析失败（PathBuf 不支持 ?）
/// 因此窗口创建交给前端，Rust 只负责数据状态，并返回 note 供前端设置初始位置尺寸
#[tauri::command]
pub fn pin_note(app: AppHandle, state: State<AppState>, id: String) -> Result<Note> {
    let note = match state.set_pinned(&id, true) {
        Ok(n) => n,
        Err(e) => {
            emit_storage_error(&app, &e);
            return Err(e);
        }
    };

    // 若窗口已存在则聚焦（前端会创建窗口，这里只处理已存在的情况）
    let label = format!("note-{}", id);
    if let Some(window) = app.get_webview_window(&label) {
        let _ = window.show();
        let _ = window.set_focus();
    }

    let _ = app.emit(event_name::NOTE_PINNED, NotePayload { id });
    Ok(note)
}

/// 收回便签（关闭窗口）
#[tauri::command]
pub fn unpin_note(app: AppHandle, state: State<AppState>, id: String) -> Result<()> {
    if let Err(e) = state.set_pinned(&id, false) {
        emit_storage_error(&app, &e);
        return Err(e);
    }

    let label = format!("note-{}", id);
    if let Some(window) = app.get_webview_window(&label) {
        let _ = window.close();
    }

    let _ = app.emit(event_name::NOTE_UNPINNED, NotePayload { id });
    Ok(())
}

/// 获取配置
#[tauri::command]
pub fn get_config(state: State<AppState>) -> Result<Config> {
    Ok(state.get_config())
}

/// 更新配置（部分字段，传 JSON 对象）
#[tauri::command]
pub fn update_config(
    app: AppHandle,
    state: State<AppState>,
    partial: serde_json::Value,
) -> Result<Config> {
    let config = match state.update_config(partial) {
        Ok(c) => c,
        Err(e) => {
            emit_storage_error(&app, &e);
            return Err(e);
        }
    };
    let _ = app.emit(event_name::CONFIG_UPDATED, ());
    Ok(config)
}

/// 导出便签到指定文件路径
#[tauri::command]
pub fn export_notes(state: State<AppState>, path: String) -> Result<()> {
    state.export_notes(&PathBuf::from(path))
}

/// 从指定文件路径导入便签
/// 导入成功后对每条新增便签 emit `note:created` 事件，通知所有窗口刷新
#[tauri::command]
pub fn import_notes(
    app: AppHandle,
    state: State<AppState>,
    path: String,
) -> Result<ImportResult> {
    let result = match state.import_notes(&PathBuf::from(path)) {
        Ok(r) => r,
        Err(e) => {
            emit_storage_error(&app, &e);
            return Err(e);
        }
    };
    // 通知前端刷新：发 imported 数量的 note:created 事件
    // 前端 stores 订阅了 note:created → refreshNotes()，收到即重新拉全量
    if result.imported > 0 {
        let _ = app.emit(event_name::NOTE_CREATED, NotePayload { id: String::new() });
    }
    Ok(result)
}

/// 获取数据目录路径（用于设置页只读显示，PRD 设置项 5）
#[tauri::command]
pub fn get_data_dir(state: State<AppState>) -> Result<String> {
    Ok(state.storage().data_dir().to_string_lossy().to_string())
}

/// 全部收回：遍历 pinned 便签，逐个 set_pinned(false) + 关闭窗口 + emit 事件（SOP 8.11）
#[tauri::command]
pub fn retract_all_notes(app: AppHandle, state: State<AppState>) -> Result<()> {
    let notes = state.list_notes();
    for note in notes.iter().filter(|n| n.pinned) {
        let _ = state.set_pinned(&note.id, false);
        let label = format!("note-{}", note.id);
        if let Some(window) = app.get_webview_window(&label) {
            let _ = window.close();
        }
        let _ = app.emit(event_name::NOTE_UNPINNED, NotePayload { id: note.id.clone() });
    }
    Ok(())
}

/// 退出应用（SOP 8.12）
#[tauri::command]
pub fn quit_app(app: AppHandle) -> Result<()> {
    app.exit(0);
    Ok(())
}

/// 重新注册全局快捷键（设置页改快捷键后调用，SOP 8.9）
/// 先 unregister_all 再 register 新 hotkey
#[tauri::command]
pub fn register_hotkey(app: AppHandle, state: State<AppState>) -> Result<()> {
    let hotkey = state.get_config().hotkey;
    app.global_shortcut()
        .unregister_all()
        .map_err(|e| Error::Shortcut(format!("取消旧快捷键失败: {}", e)))?;
    app.global_shortcut()
        .register(hotkey.as_str())
        .map_err(|e| Error::Shortcut(format!("注册快捷键 '{}' 失败: {}", hotkey, e)))?;
    Ok(())
}
